documentation/MSBuild-Server.md

```diff
@@ -0,0 +1,54 @@
+# MSBuild Server
+
+MSBuild Server is basically an another type of node which can accept build request from clients and utilize worker nodes in current fashion to build projects. Main purpose of the server node is to avoid expensive MSBuild process start during build from tools like .NET SDK.
+
+## Communication protocol
+
+The server node uses same IPC approach as current worker nodes - named pipes. This solution allows to reuse existing code. When process starts, pipe with deterministic name is opened and waiting for commands. Client has following worfklow:
+
+1. Try to connect to server
+   - If server is not running, start new instance
+   - If server is busy, fallback to classic build 
+2. Initiate handshake
+2. Issue build command with `EntryNodeCommand` packet
+3. Read packets from pipe
+   - When `EntryNodeConsoleWrite` packet type is recieved, write content to appropriate output stream with respected coloring
+   - When `EntryNodeResponse` packet type is recieved, build is done and client writes trace message with exit code
+
+### Pipe name convention & handshake
+
+There might be multiple server processes started with different architecture, associated user, MSBuild version and another options. To quickly identify the appropriate one, server uses convention that includes these options in the name of the pipe. Name has format `MSBuildServer-{hash}` where `{hash}` is a SHA256 hashed value identifying these options.
+
+Handshake is a procedure ensuring that client is connecting to a compatible server instance. 
+
+### Packets for client-server communication
+
+Server requires to introduce new packet types for IPC.
+
+`EntryNodeCommand` contains all of the information necessary for a server to run a build.
+
+| Property name            | Type                        | Description |
+|---|---|---|
+| CommandLine              | String                      | The MSBuild command line with arguments for build |
+| StartupDirectory         | String                      | The startup directory path |
+| BuildProcessEnvironment  | Dictionary<String, String>  | Environment variables for current build |
+| Culture                  | CultrueInfo                 | The culture value for current build |
+| UICulture                | CultrueInfo                 | The UI culture value for current build |
+
+`EntryNodeConsoleWrite` contains information for console output.
```

Comment:
This makes sense from the consumption side, but how are these constructed on the server?

Reply:
Using custom console writer. It's like in-proc console redirection. With redirection from the client side (as I described in another thread), we could ignore this approach.

Reply:
`Console.SetOut(TextWriter)` and `Console.SetErr(TextWriter)` is used to configure static `Console` behavior and inject own `TextWriter`. This way console would look identical even when custom tasks do not use ILogger but writ directly to `Console`.